neuralmonkey/runners/beamsearch_runner.py

```diff
@@ -1,68 +1,160 @@
+        if self._num_sessions > 1:
+            for fd in self._next_feed:
+                fd.update({self._decoder.max_steps: 0})
 
-        evaluated_bs = results[0]["bs_outputs"]
-        max_time = evaluated_bs.scores.shape[0]
+        self.result = None  # type: ExecutionResult
 
-        # pick the end of the hypothesis based on its rank
-        hyp_index = np.argpartition(
-            -evaluated_bs.scores[-1], self._rank - 1)[self._rank - 1]
-        bs_score = evaluated_bs.scores[-1][hyp_index]
+    def next_to_execute(self) -> NextExecute:
+        return (self._all_encoders,
+                {"bs_outputs": self._decoder.outputs},
+                self._next_feed)
 
-        # now backtrack
-        output_tokens = []  # type: List[str]
+    # pylint: disable=too-many-locals
+    def collect_results(self, results: List[Dict]) -> None:
+        # Recompute logits
+        # Only necessary when ensembling models
+        prev_logprobs = [res["bs_outputs"].last_search_state.prev_logprobs
+                         for res in results]
+
+        # Arithmetic mean
+        ens_logprobs = (scipy.misc.logsumexp(prev_logprobs, 0)
+                        - np.log(self._num_sessions))
+
+        # The last score is computed
+        bs_outputs = results[0]["bs_outputs"]
+        step = bs_outputs.last_dec_loop_state.step - 1
+
+        self.step += step
+        self.scores = np.append(
+            self.scores,
+            bs_outputs.last_search_step_output.scores[0:step],
+            axis=0)
+
+        self.parent_ids = np.append(
+            self.parent_ids,
+            bs_outputs.last_search_step_output.parent_ids[0:step],
+            axis=0)
+
+        self.token_ids = np.append(
+            self.token_ids,
+            bs_outputs.last_search_step_output.token_ids[0:step],
+            axis=0)
+
+        if (self._decoder.max_output_len is not None and
+                self.step > self._decoder.max_output_len):
+            self.prepare_results()
+            return
+
+        # Prepare the next feed_dict (in ensembles)
+        self._next_feed = []
+        for result in results:
+            bs_outputs = result["bs_outputs"]
+
+            search_state = bs_outputs.last_search_state._replace(
+                prev_logprobs=ens_logprobs)
+
+            fd = {self._decoder.max_steps: 1,
+                  self._decoder.search_state: search_state}
+
+            dec_feedables = bs_outputs.last_dec_loop_state
+
+            # Due to the arrays in DecoderState (prev_contexts),
+            # we have to create feed for each value separately.
+            for field in self._decoder.decoder_state._fields:
+                # We do not feed the step
+                if field == "step":
+                    continue
+                tensor = getattr(self._decoder.decoder_state, field)
+                value = getattr(dec_feedables, field)
+                if isinstance(tensor, list):
```

Comment:
takže ty decoder stejty můžou bejt jen listy ale už to nemůžou bejt listy listů ani dikty jo? Asi bych to zatim nechal, ale napsal bych sem NOTE

Reply:
Anebo elif je to tensor, a else hoď chybu. Spíš if tensor tak ho tam hoď, elif list tak hoď všechny a else mi  vynadej

Reply:
Nevim uplne presne, jak jsou implementovany ty namedtuple, kdyz k nim pristupujes, jako k polim. Kdyby v DecoderState nebyly ty arraye, tak ti staci udelat fd.update({self._decoder.decoder_state: dec_feedables}), ale zrejme to ten extra rozmer (napr. z tech prev_contexts) rozbiji.
Myslim, ze soucasne reseni je v pohode. NOTE ti muzu pridat do toho komentare nad tim.

Reply:
namedtuple je tuple, kterej navíc funguje jako namespace. k prvkům tuplu přistupuješ jako k prvkům pole, hranatejma závorkama. K prvkům napespacu přistupuješ buď přímo `obj.attr` nebo nepřímo `obj.getattr("attr")`.
namedtuple má ještě metodu `_fields`, která vrátí __pole__ stringů - atributů. Ty pak můžeš hodit tomu `getattr`.

Hoď tam pls NOTE do komentu nad tim.

Reply:
DONE